File: python/stack.py
"""*******************************************************************************
 * 소요시간: 1시간
 * 시간복잡도: O(N)
 * 메모리: 39952 kb
 * 시간: 2752 ms
 *******************************************************************************"""
from collections import deque

stack = deque()  # 스택 초기화

n = int(input())

detec_lst = []
for _ in range(n):
    a = int(input())
    detec_lst.append(a)

i = 0  # 스택에 넣을 숫자 (1부터 n까지)
result = []  # 연산 결과 (+, -) 저장

for j in range(n):
    while i < n + 1:
        if len(stack) > 0:  # 스택이 비어 있지 않을 경우
            item = stack[-1]  # 스택의 top 요소 확인
            if item == detec_lst[j]:  # 스택의 top이 목표 숫자와 일치하면 pop
                result.append("-")
                stack.pop()
                break  # 다음 숫자로 이동
            i += 1
            stack.append(i)  # 목표 숫자가 아니면 다음 숫자를 push
            result.append("+")
        else:
            i += 1  # 스택이 비어있으면 새로운 숫자를 push
            stack.append(i)
            result.append("+")

if len(stack) == 0:
    for i in range(len(result)):
        print(result[i])
else:
    print("NO")  # 스택이 비어있지 않으면 수열 불가능

File: src/stack.java
/*******************************************************************************
 * 소요시간: 30분
 * 시간복잡도: O(n)
 * 메모리: 20076 kb
 * 시간: 264 ms
 *******************************************************************************/
import java.util.Stack;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class stack {
    public static void main(String[] args) throws IOException{
        Stack<Integer> stackInt = new Stack<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int cnt = Integer.parseInt(br.readLine());

        for (int i = 0; i < cnt; i++) {
            String line = br.readLine(); // 한 줄 입력 받기
            String[] parts = line.split(" "); // 공백 기준 나눔

            String str = parts[0];

            if (str.equals("push")){
                int num = Integer.parseInt(parts[1]);
                stackInt.push(num);

            } else if (str.equals("pop")) {
                if (stackInt.isEmpty()) {
                    System.out.println(-1);
                } else {
                    System.out.println(stackInt.pop());
                }

            } else if (str.equals("top")) {
                if (stackInt.isEmpty()) {
                    System.out.println(-1);
                } else {
                    System.out.println(stackInt.peek());
                }

            } else if (str.equals("size")) {
                System.out.println(stackInt.size());

            } else if (str.equals("empty")) {
                if (stackInt.isEmpty()) {
                    System.out.println(1);
                } else {
                    System.out.println(0);
                }
            }
        }
    }
}

